server/mediaserver.go

```diff
@@ -489,12 +310,84 @@ func endRTMPStreamHandler(s *LivepeerServer) func(url *url.URL, rtmpStrm stream.
 		if monitor.Enabled {
 			monitor.LogStreamEndedEvent(cxn.nonce)
 		}
+		cxn.eof <- struct{}{}
 		delete(s.rtmpConnections, mid)
 
 		return nil
 	}
 }
 
+func (s *LivepeerServer) startSession(cxn *rtmpConnection) *BroadcastSession {
+
+	mid := rtmpManifestID(cxn.stream)
+	cpl := cxn.pl
+	var sess *BroadcastSession
+
+	// Connect to the orchestrator. If it fails, retry for as long
+	// as the RTMP stream is alive; maybe the orchestrator hasn't
+	// received the block containing the job yet
+	broadcastFunc := func() error {
+		var err error
+		sess, err = selectOrchestrator(s.LivepeerNode, cpl)
+		if err == ErrDiscovery {
+			return nil // discovery disabled, don't retry
+		} else if err != nil {
+			// Should be logged upstream
+		}
+		s.connectionLock.RLock()
+		defer s.connectionLock.RUnlock()
+		_, active := s.rtmpConnections[mid]
+		if active {
+			return err
+		}
+		return nil // stop if inactive
+	}
+	expb := backoff.NewExponentialBackOff()
+	expb.MaxInterval = BroadcastRetry
+	expb.MaxElapsedTime = 0
+	backoff.Retry(broadcastFunc, expb)
+	return sess
+}
+
+func (s *LivepeerServer) startSessionListener(cxn *rtmpConnection) {
+	mut := sync.Mutex{}
+	cv := sync.NewCond(&mut)
+	pending := false
+	runStartSession := func() {
+		sess := s.startSession(cxn)
+		mut.Lock()
+		defer mut.Unlock()
+		cxn.sess = sess
+		pending = false
+		cv.Signal()
+	}
+	glog.V(common.DEBUG).Info("Starting broadcast listener for ", cxn.pl.ManifestID())
+	for {
```

Comment:
I think this could be a buffered (size = 1) channel based alternative to using a cond var:

```
sem := make(chan struct{}, 1)

for {
    select {
        case <-cxn.needOrch:
            select {
                // If this send succeeds then no other request will be accepted until we unblock sem
                case sem <- struct{}{}: 
                    cxn.sess = s.startSession()
                    <-sem // Unblock sem
                default:
                    return // The above case is blocking so we already have an inflight request
            }
            ...
    }
}
```

Thoughts? Also, do we generally need a mutex for `cxn.sess` since it is read in a few other places?

Reply:
I like this, will give it a test. Thanks!

> Also, do we generally need a mutex for cxn.sess since it is read in a few other places?

Yes, we generally should use a mutex to protect accesses to `ctx.sess`. Needs to be added to the cxn struct though, see https://github.com/livepeer/go-livepeer/pull/646#discussion_r244882689

Reply:
Protective mutex around `cxn.sess` has been added, and tests added to this area of code. Things look clean, but will refactor around this solution shortly.

Reply:
Fixed in https://github.com/livepeer/go-livepeer/pull/646/commits/6940c486d82e6d8b0444fa5b2a7461066161ca97